2_process/src/process_nhdv2_attr.R

```diff
@@ -86,3 +110,126 @@ process_cumulative_nhdv2_attr <- function(file_path,segs_w_comids,cols){
+  # instead of the monthly values for each year
+  if(grepl("PPT_CAT",file_path)){
+    message("Calculating long-term monthly average precipitation from annual data")
+    dat <- calc_monthly_avg_ppt(dat)
+  }
+  
+  # 4. Identify columns of interest
+  # Subset VarsOfInterest table to retain the desired dataset {data_name} 
+  vars_item <- vars_table %>%
+    filter(Dataset_name == data_name) 
+  
+  col_names <- tibble(col_name = vars_item$Catchment.Item.Name,
+                      cat_agg_op = vars_item$CAT_aggregation_operation) %>%
+    filter(!is.na(col_name)) 
+  
+  # Reformat col_names for certain datasets with year suffixes on column names
+  # Function `format_col_names_years` sourced from 1_fetch/fetch_nhdv2_attributes_from_sb.R
+  # National Inventory of Dams data:
+  if(unique(vars_item$sb_id) == "58c301f2e4b0f37a93ed915a"){
+    years <- unique(str_extract(names(dat),"\\d{2,}"))
+    years <- years[!is.na(years)]
+    cols <- format_col_names_years(col_names$col_name,years,yr_pattern = "YYYY")
+    cols <- cols[cols != "COMID"]
+    col_names <- tibble(col_name = cols,
+                        cat_agg_op = rep(vars_item$CAT_aggregation_operation,length(years))) 
+  }
+  
+  # HDENS data:
+  if(unique(vars_item$sb_id) == "5910de31e4b0e541a03ac983"){
+    years <- unique(str_extract(names(dat),"\\d{2,}"))
+    years <- years[!is.na(years)]
+    cols <- format_col_names_years(col_names$col_name,years,yr_pattern = "XX")
+    cols <- cols[cols != "COMID"]
+    col_names <- tibble(col_name = cols,
+                        cat_agg_op = rep(vars_item$CAT_aggregation_operation,length(years)))
+  }
+  
+  # monthly average precipitation data:
+  if(unique(vars_item$sb_id) %in% c("5734acafe4b0dae0d5de622d")){
+    col_names$col_name <- names(dat)[names(dat) != "COMID"]
+  }
+  
+  # Separate columns based on the aggregation operation we need to scale the NHD-catchment-values to PRMS-scale values
+  cols_sum <- col_names$col_name[col_names$cat_agg_op=="sum"]
+  cols_area_wtd_mean <- col_names$col_name[col_names$cat_agg_op=="area_weighted_mean"]
+  cols_min <- col_names$col_name[col_names$cat_agg_op=="min"]
+  cols_max <- col_names$col_name[col_names$cat_agg_op=="max"]
+  
+  # 5. Munge downloaded data
+  dat_proc <- dat %>%
+    # retain CAT columns from NHDv2 attributes dataset
+    select(c(COMID,starts_with("CAT"))) %>%
+    # join data to {segs_w_comids} data frame by COMID
+    right_join(.,segs_w_comids,by=c("COMID")) %>%
+    # join data to {nhd_reaches} by COMID since we need the AREASQKM and LENGTHKM attributes not
+    # included in the downloaded NHDv2 attribute datasets
+    left_join(.,nhd_reaches,by="COMID") %>%
+    relocate("PRMS_segid",.before="COMID") 
+  
+  # Flag columns with undesired flag values (e.g. -9999)
+  flag_cols <- dat_proc %>%
+    select(where(function(x) -9999 %in% x)) %>% 
+    names()
+  
+  # For columns with undesired flag values, replace -9999 with NA, else use existing value
```

Comment:
From the metadata it looks like the following `Dataset_name`(s) from our VarsOfInterest table have this meaning (i.e., "-9999 denotes NODATA usually water"):

- STATSGO_HYDGRP
- STATSGO_TEXT
- STATSGO_LAYER

For these three datasets, I've specified that -9999's should be changed to 0 instead of NA in `process_catchment_nhdv2_attr`. The TEXT and LAYER datasets don't appear to have any rows where -9999 within our spatial extent, however, STATSGO_HYDGRP does. I don't expect this to result in substantial changes in the attr values, however I'd expect the area-weighted means to be slightly smaller, since we're now including in the calculation those COMIDs (and catchment areas) which were previously NA but are now 0.